service_container/service_locators.rst

```diff
@@ -0,0 +1,155 @@
+        {
+            if (isset($handlerMap[$commandClass = get_class($command)])) {
+                $handler = $handlerMap[$commandClass];
+
+                return $handler->handle($command);
+            }
+        }
+    }
+
+    // ...
+    $commandBus->handle(new FooCommand());
+
+Because only one command is handled at a time, other CommandHandler services are not
+used but unnecessarily instantiated.
+
+Service Locators are intended to solve this problem by giving access to a set of
+identified services while instantiating them only when really needed, providing
+laziness.
+
+Configuration
+-------------
+
+For injecting a service locator in a service, use the `service_locator` argument
+type and define the set of services that needs to be accessible from:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        services:
+
+            app.command_bus:
+                class: AppBundle\CommandBus
+                arguments:
+                    - =service_locator:
+                        AppBundle\FooCommand: '@app.command_handler.foo'
+                        AppBundle\BarCommand: '@app.command_handler.bar'
+
+    .. code-block:: xml
+
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+            <services>
+
+                <service id="app.command_bus" class="AppBundle\CommandBus">
+                    <argument type="service-locator">
+                        <argument key="AppBundle\FooCommand" type="service" id="app.command_handler.foo" />
+                        <argument key="AppBundle\BarCommand" type="service" id="app.command_handler.bar" />
+                    </argument>
+                </service>
+
+            </services>
+        </container>
+
+    .. code-block:: php
+
+        use Symfony\Component\DependencyInjection\Argument\ServiceLocatorArgument;
+        use Symfony\Component\DependencyInjection\Reference;
+
+        //...
+
+        $container
+            ->register('app.command_bus')
+            ->setArguments(array(new ServiceLocatorArgument(array(
+                'AppBundle\FooCommand' => new Reference('app.command_handler.foo'),
+                'AppBundle\BarCommand' => new Reference('app.command_handler.bar'),
+            ))))
+        ;
+
+.. note::
+
+    The services defined in the service locator argument must be keyed,
+    as keys become their unique identifier inside the locator.
+
+.. tip::
+
+    Not only services can be passed accessible through a service locator,
+    but all types which are supported by the configuration format used to
+    configure it.
+
+Usage
+-----
+
+Back to our CommandBus which would now look like::
+
+    // ...
+    use Psr\Container\ContainerInterface;
```

Comment:
Yet I understand your concerns, I think PSR-11 is legit in this context since the primary goal of the locator is to fetch services from the DIC (keeping in mind that this doc is about symfony/DI). 
I find the callable typehint useful for cases where there's no DIC behind the locator, which should not be the main usage of this feature to me since it is first built to avoid container-aware stuff inside the symfony full stack.

> semantically string here is not serviceId, but something more generic. It looks like far-fetched concept here.

In my opinion, services identifiers should remain the same in the locator than in the container from which they come from, so they're safely accessible no matter which container/registry/locator/provider is given to you, that is primordial to me.
Redefining identifiers might just be useful in some cases that I'm still not sure to get, but it seemed logic for others so here we are.

> Second example with callable looks for me better even if it has implicit interface. At least, it has nothing to do with DI-containers.

As you pointed out in the code PR, container is just a kind of map after all and if you ask me, PSR-11 could be called `Object[Store|Provider|Locator|Registry]` instead of `Container`.